src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +121,268 @@ namespace wasm {
 
 namespace {
 
+// Contains information about a RefCast we want to move to a target LocalGet.
+struct RefCastInfo {
+  LocalGet* target = nullptr;
+  RefCast* bestCast = nullptr;
+};
+
+// Contains information about a RefAs we want to move to a target LocalGet.
+// Currently only RefAsNonNull will be moved.
+struct RefAsInfo {
+  LocalGet* target = nullptr;
+  RefAs* bestCast = nullptr;
+};
+
+// Find a cast to move earlier to another local.get. More refined subtypes are
+// chosen over less refined ones.
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+  size_t numLocals;
+
+  // Tracks the current earliest local.get that we can move a cast to without
+  // side-effects, and the most refined cast that we can move to it (could be
+  // already at the earliest local.get).
+  //
+  // Note that we track a cast already on the get since we only want to move
+  // better casts to it: if the best cast is already on the get, there is no
+  // work to do.
+  std::vector<RefCastInfo> currRefCastMove;
+  std::vector<RefAsInfo> currRefAsMove;
+
+  // Used to analyze expressions to see if casts can be moved past them.
+  EffectAnalyzer testRefCast;
+  EffectAnalyzer testRefAs;
+
+  // Maps LocalGets to the most refined RefCast to move to it, to be used by the
+  // EarlyCastApplier. If the most refined RefCast is already at the desired
+  // LocalGet, it does not appear here. In the normal case, only one RefCast
+  // needs to be moved to a LocalGet. If a LocalGet is cast to multiple types
+  // which are not subtypes of each other the a trap is inevitable, and we
+  // assume this would already be optimized away beforehand, so we don't care
+  // about this special case.
+  std::unordered_map<LocalGet*, RefCast*> refCastToApply;
+
+  // Maps LocalGets to a RefAs to move to it, to be used by the
+  // EarlyCastApplier. As of right now RefAsNonNull is the only non-extern cast,
+  // so we only have one type of RefAs cast to move.
+  std::unordered_map<LocalGet*, RefAs*> refAsToApply;
+
+  EarlyCastFinder(PassOptions options, Module* module, Function* func)
+    : options(options), numLocals(func->getNumLocals()),
+      currRefCastMove(func->getNumLocals()),
+      currRefAsMove(func->getNumLocals()), testRefCast(options, *module),
+      testRefAs(options, *module) {
+
+    // TODO: generalize this when we handle more than RefAsNonNull
+    RefCast dummyRefCast(module->allocator);
+    RefAs dummyRefAs(module->allocator);
+    dummyRefAs.op = RefAsNonNull;
+
+    testRefCast.visit(&dummyRefCast);
+    testRefAs.visit(&dummyRefAs);
+  }
+
+  void flushRefCastResult(size_t index, Module& module) {
```

Comment:
Please add a comment about what this does, like
```
We track information as we go, looking for the best cast to move backwards, and when we hit a barrier - a position we can't optimize past - then we flush/finalize the work we've done so far, since nothing better can appear later.
```

Reply:
Done.